### PR TITLE
Use parse_url kernel for QUERY literal and column key

### PR DESCRIPTION
**Pull request by thirtiseven**
This PR adds plugin support for QUERY parsing in parse_url (both literal key and column key).

Contributes to https://github.com/NVIDIA/spark-rapids/issues/8963

Depends on: https://github.com/NVIDIA/spark-rapids-jni/pull/1704

Depends on: https://github.com/NVIDIA/spark-rapids-jni/pull/1719

Test results is updating in jni pr, perf test results tbd.

### CODE REVIEW

File: integration_tests/src/main/python/url_test.py
Comment (revans2) [inline integration_tests/src/main/python/url_test.py:181]: nit: It would probably be best to separate these out into different test runs, as the project will fall back to the CPU if any of them say they are not supported instead of if all of them do.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3272]: I think we might need `(` and `)` in here too for the check. 

@NVnavkumar and/or @andygrove is this a good enough check for a possible regexp problem?  Be aware that Spark is creating the regexp with a prefix and postfix

https://github.com/apache/spark/blob/fa60a7e216e63b1edb199b1610b26197815c656b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala#L132-L133

```
private val REGEXPREFIX = "(&|^)"
private val REGEXSUBFIX = "=([^&]*)"
```

And it is trying to capture the second capture group. So it is kind of a mess.

Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3272]: There is a list of regex meta characters here: https://github.com/NVIDIA/spark-rapids/blob/1acc8e0744ff411ebb7ed63e993e2a278f018236/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala#L687

I think  we should pull this out of a private val (maybe as a constant in GpuOverrides) and use it in both places.

Plus I think `-` alone could be part of a valid key to extract so there is no need to put it in a regex check here.

File: integration_tests/src/main/python/url_test.py
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py]: done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala]: done.

**Comment by hyperbolic2346** [comment]
JNI dependencies are merged now

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
Comment (NVnavkumar) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:750]: Why is the negative condition removed here? This should be negated in this part.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Thanks, done.

**Review by NVnavkumar — APPROVED** [review]

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by thirtiseven** [comment]
Found some more mismatched cases with `df.selectExpr("*", "parse_url(url, 'QUERY', 'query') as pr")` yesterday:
```
|bab.bg/en_page.php?id=33 |null |cpu     |
|bab.bg/en_page.php?id=33 |33   |gpu     |
```
Seems cpu will always return null for urls without protocol. I think it’s a small bug and we probably could let this pr merge in first and fix it next week as a bug.

@NVnavkumar help review this again thanks

**Comment by thirtiseven** [comment]
follow-up issue: https://github.com/NVIDIA/spark-rapids-jni/issues/1735

**Comment by thirtiseven** [comment]
https://github.com/NVIDIA/spark-rapids-jni/pull/1740 fixes the bug above, better to merge this after jni merged.

**Comment by thirtiseven** [comment]
JNI fix merged, it was not a break change and parse_url kernel is correct enough now. Merging this…

**Pull request merged**
